### DISCUSSION TITLE
Prefetch SSR Data

### DISCUSSION THREAD

**Discussion by fleck** [vercel/next.js#31832] [Ideas]
### Describe the feature you'd like to request

I like to add heuristic prefetching for server side rendered pages. In next.js 11 I currently accomplish prefetching ssr data like this:

```ts
<Link href={"/path-to-ssr-page"}>
  <a
    onPointerEnter={({ target }) => {
      // ... get dataHref from target
      const dataHref = Router.router?.pageLoader.getDataHref(
      // ...
      // use private method to prefetch data
      Router.router?._getServerData(dataHref)
    }}
```

### Describe the solution you'd like

A public API to prefetch data for a server side rendered page. It'd be really nice if it could take the "vanilla" url and create the dataHref for callers.

### Describe alternatives you've considered

I've pieced together a working solution using a private API, but that's subject to breaking with every next release, in fact I can't upgrade to next 12 because of some internal API changes.

Thanks for the great framework!

**Comment by neongreen** [comment]
@fleck Have you found a solution for Next 12?

**Reply by fleck** [reply]
I have not checked into a solution for next 12. I'm still on next 11 and probably won't be able to upgrade till sometime in March.

**Comment by fleck** [comment]
Hey @neongreen, had a chance to dig into this for next 12. I used https://webpack.js.org/loaders/exports-loader/ I had to tweak my next config to look like this:

```ts
module.exports = {
  webpack: (config) => {
    config.module.rules.push({
      test: require.resolve("next/dist/shared/lib/router/router.js"),
      loader: "exports-loader",
      options: {
        exports: "fetchNextData",
      },
    })

    return config
  },
}
```
Then I created a module that looks roughly like this:

```ts
import Router from "next/router"
// @ts-expect-error we're adding this export via the webpack exports-loader
import { fetchNextData } from "next/dist/shared/lib/router/router"

type Params = { [key: string]: string }

export const prefetch = ({ href, params }: { href: string; params?: Params }) => {
  const dataURL = new URL(href)

  if (params) {
    Object.entries(params).forEach(([key, value]) => {
      dataURL.searchParams.append(key, value)
    })
  }

  const dataHref = Router.router?.pageLoader.getDataHref({
    href: dataURL.toString(),
    asPath: dataURL.toString(),
    ssg: false,
    locale: Router.locale,
  })

  if (dataHref) {
    fetchNextData(dataHref, true, false, Router.router!.sdr, true)
  }
}
```

**Reply by fleck** [reply]
Hello @timneutkens, what would be the feasibility of adding a public way to enable developers to prefetch data for SSR pages? This is really useful for doing stuff the like prefetching data on mouse down or mouseover to make the UI feel more responsive.

**Reply by smontlouis** [reply]
I was looking exactly for this issue. 
We have a graphCDN on top on our graphQL server, and I need to prefetch some pages.

@fleck is your solution viable ? I might try it, thanks !


EDIT: Working flawlessly by patching next with `patch-package` by adding an export to `fetchNextData`, and then using your function. Thank you very much !

**Reply by pbell23** [reply]
Hey @bulby97 ,
Could you please share how you exported the fetchNextData function because I'm having an error when trying to call it in my code. I added a exports.fetchNextData on top of the module like the other methods. Thanks.
